rclcpp/src/rclcpp/node_interfaces/node_parameters.cpp

```diff
@@ -374,7 +374,15 @@ NodeParameters::declare_parameter(
   const rcl_interfaces::msg::ParameterDescriptor & parameter_descriptor,
   bool ignore_override)
 {
-  std::lock_guard<std::mutex> lock(mutex_);
+  std::lock_guard<std::recursive_mutex> lock(mutex_);
+
+  if (!parameter_modification_enabled_) {
+    throw rclcpp::exceptions::ParameterModifiedInCallbackException(
+            "cannot declare a parameter from within set callback");
+  }
+
+  parameter_modification_enabled_ = false;
+  RCLCPP_SCOPE_EXIT({parameter_modification_enabled_ = true;});
```

Comment:
You may want to consider moving this into a class, e.g. `ParameterMutationRecursionGuard` or something, that checks if the variable is set (throw if it is) then sets this variable to false in the constructor, and then sets it back to false in the destructor. Just to de-duplicate this code.